Contractor

Mohannad-Aldughaim opened the pull request:
Hi all.

This is WIP for using contractors in bmc as described in [here](https://arxiv.org/abs/2012.11245). 

When reviewing, I suggest starting at `goto_contractort.h`

## Goto Contractor

The class 'goto_contractort' does the following on the goto program:
-Parse the properties in the code which are inside the assert. (for now, only one assert). This will parse the expressions from ESBMC format to IBEX format.
-Parse the assumes created by `--interval-analysis` option in ESBMC. This is for creating the intervals of the variables in the assert.
-Use the Forward-backward contractor in the IBEX library to reduce the intervals.
-Insert assumes reflecting the contracted intervals.

For now, it only works correctly in one case which is if the intervals contracted are at the end of a loop.

## IBEX

-IBEX is a C++ library for constraint processing over real numbers. It has an implementation of contractors and intervals arithmetic and set operations. Can be found [here](http://ibex-lib.org/)
-The main reason to include this library in ESBMC is to handle the intervals and contractors. 

## Example

The prime example that I have been working on is a benchmark from SV-comp which is [afnp2014.c](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/loop-lit/afnp2014.c)


## Support

- Only int. As `--interval-analysis` option does not support floats.
- One property. So only one constraint in IBEX. 
- One loop. 
- No support for function calls.
- Binary arithmetic operators supported: +, -, *, / 
- Binary comparison operators supported: >,>=,<,<= 

## Implementation Notes

-Added a new option in ESBMC called `--goto-contractor`.
-Should be used with `--incremental-bmc` for now.

Mohannad-Aldughaim committed:
Added goto_contractor class
Added Ibex to cmake files

Mohannad-Aldughaim committed:
Added goto-contractor option to options and help
Added Contractor code

Mohannad-Aldughaim committed:
initial conversion of expressions

Mohannad-Aldughaim committed:
Added goto_expr_factory.cpp and goto_expr_factory.h to create expressions for assumes.
Fixed association with k-induction and now works with afnp2014.c

Mohannad-Aldughaim committed:
Added goto_expr_factory.cpp and goto_expr_factory.h to create expressions for assumes.
Fixed association with k-induction and now works with afnp2014.c

Mohannad-Aldughaim committed:
Clean up for code and support for more expressions

Mohannad-Aldughaim committed:
Clean up for code and support for more expressions

Mohannad-Aldughaim committed:
Added option for goto_contractor.
rebuilt interval parsing.

mikhailramalho commented on src/esbmc/bmc.cpp:
```diff
@@ -421,7 +421,7 @@ void bmct::bidirectional_search(
   // in k-induction mode
   if(!(options.get_bool_option("inductive-step") &&
-       options.get_bool_option("k-induction")))
+       options.get_bool_option("k-induction") ))
```
remove extra space.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1580,6 +1580,12 @@ bool esbmc_parseoptionst::process_goto_program(
+    {
+      interval_analysis(goto_functions, ns);
+      goto_contractort(goto_functions, msg);
+    }
```
this will run the interval analysis twice if the `--interval-analysis` is used with the new flag.

You can change:
```
    if(cmdline.isset("interval-analysis"))
      interval_analysis(goto_functions, ns);
```
to
```
    if(cmdline.isset("interval-analysis") || cmdline.isset("goto-contractor"))
      interval_analysis(goto_functions, ns);
```
and remove the interval analysis from this branch.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+
+#include "goto_contractor.h"
+
+using namespace std;
```
Remove this and use the `std` namespace explicitly.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+{
+  //only main here
+  auto function = goto_functions.function_map.find("c:@F@main");
+  std::cout << "\n parsing assumes\nfound main: " << function->first << "\n";
```
You need to remove the cout and use our messaget interface for messages.

Mohannad-Aldughaim replied:
Replaced all cout with messaget interface.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+  std::shared_ptr<relation_data> rel;
+  rel = dynamic_pointer_cast<relation_data>(expr);
+
+  //expr->dump();
```
remove unnecessary comments.

Mohannad-Aldughaim replied:
done. Though I left some for reference later. Will be removed when I don't need them anymore.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+  //side 2 is always a number.
+  auto side2 = rel->side_2;
+
+  if(is_typecast2t(side1))
```
You can use `get_base_object` here, then check if the base object is a symbolt. If it isn't return.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+
+  if(!is_constant_int2t(side2) && !is_constant_floatbv2t(side2))
+    return;
+  if(is_constant_int2t(side2))
```
`value` is a double. You'll probably need to handle the values differently here.

Mohannad-Aldughaim replied:
This one is a bit complex. The library Ibex uses doubles in intervals. So I need to convert from int to double and back to int. Not sure yet of the safest way to do this.

mikhailramalho replied:
Using BigInt should do the trick... When you need to convert to double just call .toint64().

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+
+  int index =
+    map->find((string)symbol->get_symbol_name().c_str());
+  if(index == -1) return;
```
move return to the next line.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+    return;
+
+  int index =
+    map->find((string)symbol->get_symbol_name().c_str());
```
you probably don't need this cast.

Mohannad-Aldughaim replied:
removed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+  int index =
+    map->find((string)symbol->get_symbol_name().c_str());
+  if(index == -1) return;
+  if(is_greaterthan2t(expr) || is_greaterthanequal2t(expr))
```
You can replace this with a switch over the expr type.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+  goto_functionst goto_functions,
+  IntervalVector new_intervals)
+{
+  std::cout << "Inserting assumes.. " << std::endl;
```
use messaget interface.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+      isfinite(new_intervals[i].lb()) &&
+      map->intervals[i].lb() != new_intervals[i].lb())
+    {
+      if(is_signedbv_type(X))
```
remove empty if.

Mohannad-Aldughaim replied:
removed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,359 @@
+IntervalVector goto_contractort::contractor()
+{
+  NumConstraint c = *constraint;
+  std::cout << "My Constraint:" << c << std::endl;
```
use messaget interface.

Mohannad-Aldughaim replied:
done.

mikhailramalho reviewed (changes requested):
I've seen the first few changes and there are already several changes needed. Please, fix them and I'll take another look.

lucasccordeiro commented:
@Mohannad-Aldughaim: @mikhailramalho's comments are very relevant here. Could you please address them before others can review your PR? Additionally, there are conflicting files in your PR. Could you please solve them?

Mohannad-Aldughaim commented:
Will do. Thanks for the comments.

Mohannad-Aldughaim committed:
Addressed some comments from the draft pull request.

Mohannad-Aldughaim committed:
Changed output message from cout to message interface.
Resolved the conflicts from main repository.

Mohannad-Aldughaim committed:
Completely removed cout.
Changed value from double to long. However, this will need to be discussed.

Mohannad-Aldughaim committed:
removed if else chain and replaced it with switch over expression id.
support scalars in assert.

lucasccordeiro commented on src/util/goto_expr_factory.cpp:
```diff
@@ -0,0 +1,323 @@
+// Rafael Sá Menezes - 04/2020
```
Should this header be updated?

lucasccordeiro commented on src/util/goto_expr_factory.cpp:
```diff
@@ -0,0 +1,323 @@
+
+#include "goto_expr_factory.h"
+
+bool is_symbols_equal(expr2tc &v1, expr2tc &v2)
```
Can I ask you to write some short documentation for each method?

lucasccordeiro commented on src/util/goto_expr_factory.cpp:
```diff
@@ -0,0 +1,323 @@
+  return value1->value.compare(value2->value) == 0;
+}
+
+symbol2tc create_unsigned_32_symbol_expr(std::string name)
```
Why do you need 32? What about the other bit-widths? 16, 64, 128?

rafaelsamenezes replied:
added a second parameter to work with any type

lucasccordeiro commented on src/util/goto_expr_factory.cpp:
```diff
@@ -0,0 +1,323 @@
+
+greaterthanequal2tc create_greaterthanequal_relation(expr2tc &lhs, expr2tc &rhs)
+{
+  typecast2tc lhs_typecast (get_int32_type(),lhs);
```
What about other bit-widths? 16, 64, 128?

lucasccordeiro commented on src/util/goto_expr_factory.h:
```diff
@@ -0,0 +1,98 @@
+/*******************************************************************\
+Module: Cache unit test utilities
+Author: Rafael Sá Menezes
```
Can you please update this header?

lucasccordeiro commented on src/solvers/smt/smt_conv.cpp:
```diff
@@ -942,7 +942,12 @@ smt_astt smt_convt::convert_ast(const expr2tc &expr)
     else
     {
-      assert(is_signedbv_type(lte.side_1) && is_signedbv_type(lte.side_2));
+      if(!(is_signedbv_type(lte.side_1) && is_signedbv_type(lte.side_2)))
```
Should this piece of code be considered as debug? If so, add something like #ifdef DEBUG .... #endif

mikhailramalho replied:
yes, this needs to go away.

Mohannad-Aldughaim replied:
reverted to original file

lucasccordeiro commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,168 @@
+
+  IntervalVector contractor();
+
+  ibex::CmpOp get_complement(ibex::CmpOp);
```
Could you please write a short documentation for each method?

Mohannad-Aldughaim replied:
done

lucasccordeiro commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,168 @@
+      //initialize map
+      map = new MyMap();
+      vars = new Variable(MAX_VAR);
+      //find properties
```
These comments are poor.

Mohannad-Aldughaim replied:
changed to Doxygen format

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,430 @@
+{
+  std::string var_name = to_symbol2t(expr).get_symbol_name().c_str();
+  int index = map->find(var_name);
+  if(index == -1)
```
Why index == -1?

Mohannad-Aldughaim replied:
changed to NOT_FOUND value to indicate that the variable does not exist in the map.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,430 @@
+  auto it = goto_functions.function_map.find("c:@F@main");
+  auto goto_function = it->second;
+
+  for(int i = 0; i < map->size(); i++)
```
should `size_t` be used here instead of `int`?

Mohannad-Aldughaim replied:
changed.

Mohannad-Aldughaim committed:
Added documentation to get_constraints, get_intervals and contractor functions.

Mohannad-Aldughaim committed:
Added documentation to the rest of the functions.

Mohannad-Aldughaim committed:
Fixed a bug when dealing with negative numbers.
Added support for && in assumes when parsing intervals.

rafaelsamenezes committed:
[goto-contractor] ibex is now optional

rafaelsamenezes committed:
[options] reverted changes

rafaelsamenezes committed:
[goto-contractor] simplified generation library

mikhailramalho commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1584,6 +1590,15 @@ bool esbmc_parseoptionst::process_goto_program(
+      #else
+      msg.error("Current build does not support contractors");
+      #endif
+    }
```
define indentation needs to be fixed here

Mohannad-Aldughaim replied:
fixed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+  {
+    if(it->is_assume())
+    {
+      it->dump();
```
remove dump

Mohannad-Aldughaim replied:
removed

Mohannad-Aldughaim committed:
Added regression folder that includes 32 test cases for each operator.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+  auto side2 = rel->side_2;
+
+  auto obj = get_base_object(side1);
+  if(is_symbol2t(obj))
```
Can you do `if(!is_symbol(obj)) return;`?

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+    symbol = to_symbol2t(obj);
+  else
+    return;
+  ///keeping this for later
```
Do we really need to keep this?

Mohannad-Aldughaim replied:
Not any more. Removed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+  }
+  if(!is_constant_int2t(side2) && !is_constant_floatbv2t(side2))
+    return;
+  if(is_constant_int2t(side2))
```
Can you do `if(!is_constant_int2t(side2))` return;?

Mohannad-Aldughaim replied:
changed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+
+  int index = map->find(symbol->get_symbol_name());
+  if(index == NOT_FOUND)
+    return;
```
what's not_found here?

Mohannad-Aldughaim replied:
It is defined to indicate that this symbol is not in the tracked symbols list. Its value is -1.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+}
+
+IntervalVector goto_contractort::contractor()
+{
```
remove the unnecessary comments here

Mohannad-Aldughaim replied:
removed

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+  switch(expr->expr_id)
+  {
+  case expr2t::expr_ids::greaterthanequal_id:
+    c = new NumConstraint(*vars, (*f)(*vars) >= (*g)(*vars));
```
do we call delete on these object anywhere?

Mohannad-Aldughaim replied:
Created destructor.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+  default:
+    break;
+  }
+  ///keeping this for later
```
do we really need to keep this?

Mohannad-Aldughaim replied:
Not anymore. Removed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+
+  if(is_comp_expr(expr))
+  {
+    //Abort contractor
```
?

Mohannad-Aldughaim replied:
Updated to return nullptr so we can resume verification if goto-contractor is not applicable.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,453 @@
+  return index;
+}
+
+bool goto_contractort::run1()
```
can you rename this function to something that's more meaningful?

Mohannad-Aldughaim replied:
Done.

mikhailramalho reviewed (commented):
Overall I think it needs more cleanups but it seems like it's getting there.

rafaelsamenezes committed:
[goto-contractor] refactored factory calls

rafaelsamenezes committed:
[goto-contractor] cmake now searches for ibex library

Mohannad-Aldughaim committed:
[goto-contractor]
-Addressed Mikhail comments in PR 650
-Added a way to continue verification if contractor fails.
-Added destructor to free allocated objects along with lists to keep track.
-Fixed some includes.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,338 @@
+  }
+  if(!is_constant_int2t(side2))
+    return;
+  if(is_constant_int2t(side2))
```
At this point we know that side2 is a constant_int2tc, so no need to check it again.

Mohannad-Aldughaim replied:
> Using BigInt should do the trick... When you need to convert to double just call .toint64(). 

May I ask you to clarify? i didnt understand what you mean by using BigInt.

mikhailramalho replied:
BigInt is a type that can hold large integers and that can be converted to `int64_t` and `uin64_t`. Since you don't seem to be handling fp intervals, you can make `value` a `BigInt`, then call `value.to_int64()` when calling the `ibex` functions.

Mohannad-Aldughaim replied:
changed.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,338 @@
+    map->update_ub_interval(value, index);
+    break;
+  default:;
+    //error
```
you can use the message interface and call abort();

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,239 @@
+      symbols[n] = symbol;
+      var_name[n] = name;
+
+      ///keep for later. set initial intervals
```
remove comment.

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,239 @@
+   * and NumConstraint
+   */
+  ~goto_contractort()
+  {
```
can you do something like:

for( auto c : list)
  delete(p);
list.clear();

I don't think you need to check if p is not a nullptr before calling the destructor, at least you don't need it for free.

mikhailramalho reviewed (commented):
Only minor comments now.

Please before the next review check the code style:
1. One line ifs/for should not use curly brackets.
2. Remove unnecessary comments.
3. Check code-style with the same clang-format we use in the CI.

mikhailramalho commented:
Also, please look at lucas' comment about a debug code.

Mohannad-Aldughaim committed:
[goto-contractor]
-Changed types of iterator to size_t
-Disabled destructor for now.
-Removed redundant checks for NOT_FOUND
-Added monotonicity check, However, this is incomplete.

fbrausse commented on scripts/cmake/FindIbex.cmake:
```diff
@@ -0,0 +1,37 @@
+
+if(IBEX_INCLUDE_DIRS STREQUAL "IBEX_INCLUDE_DIRS-NOTFOUND")
+        message(FATAL_ERROR "Could not find ibex include headers, please check IBEX_DIR")
+    endif()
```
Please fix the indentation here.

fbrausse commented on src/esbmc/options.cpp:
```diff
@@ -236,6 +236,7 @@ const struct group_opt_templ all_cmd_options[] = {
     {"forward-condition", NULL, "check the forward condition"},
     {"inductive-step", NULL, "check the inductive step"},
     {"k-induction", NULL, "prove by k-induction "},
+    {"goto-contractor", NULL, "prove by using contractors"},
```
Could we be a bit more precise here? Technically, contractors do not prove by themselves, just in concert with other techniques. I would suggest to label this option something akin "enable contractor-based interval refinements on goto level".

Mohannad-Aldughaim replied:
I agree. changed

fbrausse commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,384 @@
+// Created by Mohannad Aldughaim on 09/01/2022.
+//
+
+#include "goto_contractor.h"
```
Please use non-local #includes for uniformity, e.g. `#include <goto-programs/goto_contractor.h>`. The `src` subdirectory is an include-directory by default.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+#ifndef ESBMC_GOTO_CONTRACTOR_H
+#define ESBMC_GOTO_CONTRACTOR_H
+
+#include "goto_k_induction.h"
```
Please use non-local #includes everywhere.

Mohannad-Aldughaim replied:
changed

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+#define MAX_VAR 10
+#define NOT_FOUND -1
+
+using namespace ibex;
```
Please don't pollute the global namespace in headers by `using namespace`; just prefix the corresponding symbols with `ibex::`.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+    if(function_loops.size())
+    {
+      map = new CspMap();
+      vars = new Variable(MAX_VAR);
```
Couldn't `map` and `vars` be plain sub-objects instead of being pointers to heap-allocated objects? Same question for other members of this class.

Mohannad-Aldughaim replied:
changed CspMap. But vars remained because it is used with every Function object created in ibex.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+  ibex::NumConstraint *constraint;
+
+  std::list<Function *> list_of_functions;
+  std::list<NumConstraint *> list_of_constraints;
```
Could we instead of a list of pointers use lists of objects, e.g. `std::list<ibex::Function>`?

fbrausse commented on src/solvers/smt/smt_conv.cpp:
```diff
@@ -268,8 +268,8 @@ smt_astt smt_convt::convert_ast(const expr2tc &expr)
-    expr->foreach_operand(
-      [this, &args, &i](const expr2tc &e) { args[i++] = convert_ast(e); });
+    expr->foreach_operand([this, &args, &i](const expr2tc &e)
+                          { args[i++] = convert_ast(e); });
```
This formatting change is caused by a new version of clang, use clang-11 to get the original formatting.

Mohannad-Aldughaim replied:
changed to clang-11

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1584,6 +1590,16 @@ bool esbmc_parseoptionst::process_goto_program(
+    if(cmdline.isset("goto-contractor"))
+    {
+#ifdef ENABLE_GOTO_CONTRACTOR
+      goto_contractort(goto_functions, msg);
```
Creating an anonymous temporary object of type `goto_contractort` is a perfectly legal way to run this functionality, but it is somewhat unconventional inside ESBMC. Would you consider making this (maybe including the #ifdef guards) a call to a proper function?

Mohannad-Aldughaim replied:
May I ask you to clarify? I have never used such a thing before.

fbrausse replied:
Well, you have done it here :)

What happens here is a call to the constructor `goto_contractort(...)`. This call creates an "anonymous" and in this case also "temporary" object, because it not assigned to anything. Here it is equivalent to code that would create a named object like so: `goto_contractort var(...)` and there is no difference between the two, because the the lifetime of this `var` object directly ends on the next line as does that of the temporary version.

So it runs constructor followed by destructor, ok, fine; but it requires a bit more thinking (at least for me) to understand what's going on. My suggestion was to wrap this into a stand-alone function and call that, maybe like so:
```c++
void goto_contractor(...)
{
#ifdef ENABLE_GOTO_CONTRACTOR
  goto_contractort(...); /* all functionality invoked by constructor */
#else
  msg.error(...);
  abort();
#endif
}
```
It's just a suggestion; it might already be enough to just add such a comment in order to minimize confusion. The advantage of having such a function is that you could also move the invocation of the algorithm (labelled by messages "1/5", "2/5", etc.) from the constructor to this function and would have a more object-oriented approach in that external algorithms could just work with an object of type `goto_contractort` instead of embedding the algorithm inside it.

Mohannad-Aldughaim replied:
Thanks for the feedback. I have changed it now to work like other options in ESBMC. Particularly, similar to k-induction where there's a function call called goto_contractor and inside it will crate the object goto_contractort. Also added the #ifdef guards.

fbrausse commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -1584,6 +1590,16 @@ bool esbmc_parseoptionst::process_goto_program(
+#ifdef ENABLE_GOTO_CONTRACTOR
+      goto_contractort(goto_functions, msg);
+#else
+      msg.error("Current build does not support contractors");
```
Please include a hint to the user how the build process should be altered to support contractors, referring to the corresponding CMake variables.

fbrausse commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -0,0 +1,384 @@
+{
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+  while(it != function->second.body.instructions.end())
```
This loop and the one in the function below could be simplified by `for (const auto &ins : function->second.body.instruction) { ... }`.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+
+  Ctc *create_contractors_from_expr2t(irep_container<expr2t>);
+  NumConstraint *create_constraint_from_expr2t(irep_container<expr2t>);
+  Function *create_function_from_expr2t(irep_container<expr2t>);
```
Instead of a heap-allocated `Function` this function could return a [`std::optional<Function>`](https://en.cppreference.com/w/cpp/utility/optional) instead. Same for `create_constraint_from_expr2t` above. This way we would reduce the number of heap allocations and indirections in accessing the objects.

Mohannad-Aldughaim replied:
I tried to do it using normal objects but ibex was not copying the values. so I kept it as pointers, for now, to be able to check if any of them failed by checking for nullptr.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+#include "util/type_byte_size.h"
+
+#define MAX_VAR 10
+#define NOT_FOUND -1
```
Please consider making these `static constexpr` members of the `goto_contractort` class since they are specific to that functionality.

Mohannad-Aldughaim replied:
changed.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -0,0 +1,257 @@
+  int find(std::string name)
+  {
+    for(size_t i = 0; i < n; i++)
+      if(var_name[i] == name)
```
Isn't it possible to use a proper `std::map` or even `std::unordered_map` instead of this linear search using string comparisons? That way you might also be able to get rid of the rather arbitrary `MAX_VAR` limit.

Mohannad-Aldughaim replied:
done

fbrausse reviewed (commented):
I've added a few remarks and suggestions. Overall it seems there still are some things to do, but it is taking shape, nice!

Mohannad-Aldughaim committed:
[goto-contractor]
-Reverted smt_conv.cpp to original file.
-Changed formatting to clang-11.
-Changed pointers to shared pointers.
-Removed destructor.
-Changed all includes to non-local.
-Removed namespace pollution and renamed each symbol with the prefix ibex::
-Removed indentations in file find_ibex.cmake
-Changed option description.
-Changed MAX_VAR and NOT_FOUND to static constexpr.
-Added a hint on how to build ESBMC with ibex library.

Mohannad-Aldughaim committed:
[goto-contractor]
-Removed shared pointers because it was causing crashes. (ibex deletes them)
-Used iterators as recommended by Franz.
-Fixed issue with arithmatic operations.
-removed unnecessary comments

Mohannad-Aldughaim committed:
[goto-contractor]
-Changed the call in esbmc_parseoptions.cpp from creating an anonymous object to function call that will instantiate and object to run the algorithm.
-In file goto_contractor.cpp, changed type of variable value in function goto_contractort::parse_intervals to BigInt.
-Changed the implementation of CspMap class to utilize std::map instead of arrays and arbitrary max size.
-Removed unnecessary code.
-Changed the variable map from a pointer to a heap allocated object to plain sub-objects.
-Added class vart to represent each variable. It contains name, index(for use with ibex), symbol and interval.

rafaelsamenezes committed:
code-style fix

Mohannad-Aldughaim committed:
[goto-contractor]
-Edit cmake file to only include when ibex is found.
-Changed return type of contractor function.

Mohannad-Aldughaim committed:
[goto-contractor]
-Lint

rafaelsamenezes committed:
Update build.yml

mikhailramalho reviewed (approved):
LGTM! As long as the CI is green, I think we are good to go.

Mohannad-Aldughaim committed:
-removed python in CI

Mohannad-Aldughaim committed:
-changed ibex source

Mohannad-Aldughaim committed:
-changed install directory for ibex

Mohannad-Aldughaim committed:
-reverted install directory for ibex

Mohannad-Aldughaim committed:
-Changed installation option of ibex to --destdir=$PWD/ibex

Mohannad-Aldughaim committed:
-Changed installation option of ibex to --destdir=$PWD/../ibex

Mohannad-Aldughaim committed:
-Changed installation option of ibex to sudo ./waf install
